File: extract_fees.php
<?php
// extract_fees.php
$html_files = glob('Fees_Chart/*.html');
$fees_data = [];

if (!is_dir('data')) {
    mkdir('data', 0777, true);
}

foreach ($html_files as $filepath) {
    $filename = basename($filepath);
    $uni_key = str_replace(['-fees.html', '.html'], '', $filename);
    
    $html = file_get_contents($filepath);
    
    // Suppress warnings for malformed HTML
    libxml_use_internal_errors(true);
    $dom = new DOMDocument();
    $dom->loadHTML(mb_convert_encoding($html, 'HTML-ENTITIES', 'UTF-8'));
    libxml_clear_errors();
    
    $xpath = new DOMXPath($dom);
    
    // Extract root CSS variables
    $theme = [
        'theme_color' => '#0066CC',
        'theme_dark' => '#004C99',
        'theme_accent' => '#231F20',
        'theme_light' => '#e0f2fe'
    ];
    $style_nodes = $xpath->query('//style');
    if ($style_nodes->length > 0) {
        $style_text = $style_nodes->item(0)->nodeValue;
        $lines = explode("\n", $style_text);
        foreach ($lines as $line) {
            $line = trim($line);
            if (strpos($line, '--theme-color:') === 0) $theme['theme_color'] = trim(explode(';', explode(':', $line)[1])[0]);
            if (strpos($line, '--theme-dark:') === 0) $theme['theme_dark'] = trim(explode(';', explode(':', $line)[1])[0]);
            if (strpos($line, '--theme-accent:') === 0) $theme['theme_accent'] = trim(explode(';', explode(':', $line)[1])[0]);
            if (strpos($line, '--theme-light:') === 0) $theme['theme_light'] = trim(explode(';', explode(':', $line)[1])[0]);
        }
    }

    // Extract Header Text
    $h1 = ''; $h2 = ''; $p_desc = '';
    $header_texts = $xpath->query('//div[contains(@class, "header-text")]');
    if ($header_texts->length > 0) {
        $header_div = $header_texts->item(0);
        $h1_node = $xpath->query('.//h1', $header_div);
        if ($h1_node->length > 0) $h1 = trim($h1_node->item(0)->nodeValue);
        $h2_node = $xpath->query('.//h2', $header_div);
        if ($h2_node->length > 0) $h2 = trim($h2_node->item(0)->nodeValue);
        $p_node = $xpath->query('.//p', $header_div);
        if ($p_node->length > 0) $p_desc = trim($p_node->item(0)->nodeValue);
    }

    // Extract Logo
    $logo_src = '';
    $logos = $xpath->query('//img[contains(@class, "static-logo")]');
    if ($logos->length > 0) {
        $logo_src = $logos->item(0)->getAttribute('src');
    }

    // Extract Categories and Tables
    $categories = [];
    $accordions = $xpath->query('//button[contains(@class, "accordion")]');
    
    foreach ($accordions as $acc) {
        $cat_name = trim($acc->nodeValue);
        
        // Find the next div with class "panel"
        $panel = $acc->nextSibling;
        while ($panel && ($panel->nodeType !== XML_ELEMENT_NODE || strpos($panel->getAttribute('class'), 'panel') === false)) {
            $panel = $panel->nextSibling;
        }
        
        if (!$panel) continue;
        
        $table_data = ['headers' => [], 'rows' => []];
        $tables = $xpath->query('.//table', $panel);
        if ($tables->length > 0) {
            $table = $tables->item(0);
            $trs = $xpath->query('.//tr', $table);
            if ($trs->length > 0) {
                // Headers
                $ths = $xpath->query('.//th', $trs->item(0));
                if ($ths->length == 0) $ths = $xpath->query('.//td', $trs->item(0));
                foreach ($ths as $th) {
                    $table_data['headers'][] = trim($th->nodeValue);
                }
                
                // Rows
                for ($i = 1; $i < $trs->length; $i++) {
                    $row = [];
                    $tds = $xpath->query('.//td', $trs->item($i));
                    foreach ($tds as $td) {
                        $row[] = trim($td->nodeValue);
                    }
                    $table_data['rows'][] = $row;
                }
            }
        }
        
        // Extract Note / PDF link
        $pdf_link = '';
        $remarks = '';
        $notes = $xpath->query('.//div[contains(@class, "note")]', $panel);
        if ($notes->length > 0) {
            $note_div = $notes->item(0);
            
            // Extract download bar
            $dl_bars = $xpath->query('.//div[contains(@class, "download-bar")]', $note_div);
            if ($dl_bars->length > 0) {
                $dl_bar = $dl_bars->item(0);
                $onclick = $dl_bar->getAttribute('onclick');
                if (preg_match("/downloadPDF\(['\"]([^'\"]+)['\"]\)/", $onclick, $matches)) {
                    $pdf_link = $matches[1];
                }
                // Remove download bar from DOM to get just remarks
                $dl_bar->parentNode->removeChild($dl_bar);
            }
            
            // Get innerHTML of note
            $remarks = '';
            foreach ($note_div->childNodes as $child) {
                $remarks .= $dom->saveHTML($child);
            }
            $remarks = trim($remarks);
        }
        
        $categories[] = [
            'category_name' => $cat_name,
            'table' => $table_data,
            'pdf_link' => $pdf_link,
            'remarks' => $remarks
        ];
    }
    
    $fees_data[$uni_key] = [
        'id' => $uni_key,
        'title' => $h1,
        'subtitle' => $h2,
        'validity' => $p_desc,
        'logo' => $logo_src,
        'theme' => $theme,
        'categories' => $categories
    ];
}

file_put_contents('data/fees_data.json', json_encode($fees_data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
echo "Extraction complete. data/fees_data.json created.\n";
?>

File: extract_fees.py
import os
import glob
import json
from bs4 import BeautifulSoup
import re

html_files = glob.glob('Fees_Chart/*.html')
fees_data = {}

for filepath in html_files:
    filename = os.path.basename(filepath)
    uni_key = filename.replace('-fees.html', '').replace('.html', '')
    
    with open(filepath, 'r', encoding='utf-8') as f:
        soup = BeautifulSoup(f, 'html.parser')
    
    # Extract root CSS variables (Theme Colors)
    style_tag = soup.find('style')
    theme = {
        'theme_color': '#0066CC',
        'theme_dark': '#004C99',
        'theme_accent': '#231F20',
        'theme_light': '#e0f2fe'
    }
    if style_tag:
        style_text = style_tag.text
        for line in style_text.split('\n'):
            line = line.strip()
            if line.startswith('--theme-color:'): theme['theme_color'] = line.split(':')[1].split(';')[0].strip()
            elif line.startswith('--theme-dark:'): theme['theme_dark'] = line.split(':')[1].split(';')[0].strip()
            elif line.startswith('--theme-accent:'): theme['theme_accent'] = line.split(':')[1].split(';')[0].strip()
            elif line.startswith('--theme-light:'): theme['theme_light'] = line.split(':')[1].split(';')[0].strip()

    # Extract Header Text
    header_text_div = soup.find('div', class_='header-text')
    h1 = header_text_div.find('h1').text.strip() if header_text_div and header_text_div.find('h1') else ''
    h2 = header_text_div.find('h2').text.strip() if header_text_div and header_text_div.find('h2') else ''
    p_desc = header_text_div.find('p').text.strip() if header_text_div and header_text_div.find('p') else ''

    # Extract Logo
    logo_img = soup.find('img', class_='static-logo')
    logo_src = logo_img['src'] if logo_img and logo_img.has_attr('src') else ''

    # Extract Categories and Tables
    categories = []
    accordions = soup.find_all('button', class_='accordion')
    for acc in accordions:
        cat_name = acc.text.strip()
        panel = acc.find_next_sibling('div', class_='panel')
        if not panel: continue
        
        table = panel.find('table')
        table_data = {'headers': [], 'rows': []}
        if table:
            trs = table.find_all('tr')
            if trs:
                # First row is headers
                th_elements = trs[0].find_all(['th', 'td'])
                table_data['headers'] = [th.text.strip() for th in th_elements]
                
                # Rest are rows
                for tr in trs[1:]:
                    tds = tr.find_all('td')
                    table_data['rows'].append([td.text.strip() for td in tds])

        # Extract Notes and PDF link
        note_div = panel.find('div', class_='note')
        pdf_link = ''
        remarks = ''
        if note_div:
            # find download_bar
            dl_bar = note_div.find('div', class_='download-bar')
            if dl_bar and dl_bar.has_attr('onclick'):
                match = re.search(r"downloadPDF\(['\"]([^'\"]+)['\"]\)", dl_bar['onclick'])
                if match:
                    pdf_link = match.group(1)
                dl_bar.extract() # remove it from note text
                
            # clean up remaining note text (convert <br> to \n for JSON, or keep HTML)
            # Keeping HTML is safer to preserve formatting
            remarks = note_div.decode_contents().strip()

        categories.append({
            'category_name': cat_name,
            'table': table_data,
            'pdf_link': pdf_link,
            'remarks': remarks
        })

    fees_data[uni_key] = {
        'id': uni_key,
        'title': h1,
        'subtitle': h2,
        'validity': p_desc,
        'logo': logo_src,
        'theme': theme,
        'categories': categories
    }

with open('data/fees_data.json', 'w', encoding='utf-8') as out:
    json.dump(fees_data, out, indent=2, ensure_ascii=False)

print("Extraction complete. data/fees_data.json created.")
